emission/analysis/modelling/trip_model/gradient_boosted_decision_tree.py

```diff
@@ -0,0 +1,122 @@
+
+        Replacement modes are considered to be the second-best choice for
+        a given trip (i.e., what mode would have been chosen if the actual
+        choice wasn't available).
+
+        The model is currently trained on data from all users.
+        """
+        if config is None:
+            config = eamtc.get_config_value_or_raise('model_parameters.gbdt')
+            logging.debug(f'GradientBoostedDecisionTree loaded model config from file')
+        else:
+            logging.debug(f'GradientBoostedDecisionTree using model config argument')
+        expected_keys = [
+            'incremental_evaluation',
+            'feature_list',
+            'dependent_var'
+        ]
+        for k in expected_keys:
+            if config.get(k) is None:
+                msg = f"gbdt trip model config missing expected key {k}"
+                raise KeyError(msg)
+        self.is_incremental = config['incremental_evaluation']
+        # Use the sklearn implementation of a GBDT
+        self.gbdt = GradientBoostingClassifier(n_estimators=50)
+        # Which features to use in the fit/prediction
+        self.feature_list = config['feature_list']
+        self.dependent_var = config['dependent_var']
+
+    def fit(self, trips: List[ecwc.Confirmedtrip]):
+        """train the model by passing data, where each row in the data
+        corresponds to a label at the matching index of the label input
+
+        :param trips: 2D array of features to train from
+        """
+        logging.debug(f'fit called with {len(trips)} trips')
+        unlabeled = list(filter(lambda t: len(t['data']['user_input']) == 0, trips))
+        if len(unlabeled) > 0:
+            msg = f'model.fit cannot be called with unlabeled trips, found {len(unlabeled)}'
+            raise Exception(msg)
+        X_train, y_train = self.extract_features(trips)
+        self.gbdt.fit(X_train, y_train)
+        logging.info(f"gradient boosted decision tree model fit to {len(X_train)} rows of trip data")
+
+    def predict(self, trip: ecwc.Confirmedtrip) -> Tuple[List[Dict], int]:
+        logging.debug(f"running gradient boosted mode prediction")
+        X_train, y_train = self.extract_features(trip)
+        y_pred = self.gbdt.predict(X_train)
+        if y_pred is None:
+            logging.debug(f"unable to predict bin for trip {trip}")
+            return []
+        else:
+            logging.debug(f"made predictions {y_pred}")
+            return y_pred
+
+    def to_dict(self) -> Dict:
+        return self.gbdt
+
+    def from_dict(self, model: Dict):
+        self.gbdt = model
+
+    def extract_features(self, trips: ecwc.Confirmedtrip) -> List[float]:
+        # TODO: need to enable generic paths other than just user input for features
+        X = pd.DataFrame(
+            [[trip['data']['user_input'][x] for x in self.feature_list] for trip in trips],
+            columns=self.feature_list
+        )
```

Comment:
Aren't some (most?) of your input features demographic? How can you read them from the user input in that case?

Reply:
This seems to be unchanged; I assume you will tackle it next?